### DISCUSSION TITLE
Using str() to localize expressions

### DISCUSSION THREAD

**Discussion by jeromecambon** [lit/lit#2270] [Q&A] [answered]
Hello,

I'm trying to localize a property, using the str() function as specified in the doc:

```
<label>${msg(str`${this.label}`)}</label>
```

The label property can take many values, but I want to localize the default one ("Default label").

This default label string is made localizable:
```
  @property({ attribute: 'label', type: String, reflect: true })
  label: string = msg('Default label');
```

So extracting and building the localization files, I get the the following 'fr.ts' (for the french translation):
```
    import {str} from '@lit/localize';
    export const templates = {
      'sd17994d58fd80ca0': `Label par défaut`,
      'saf63d34c8601dd41': str`${0}`,
    };
```

I would expect that my html label is translated when the lang is changed, but it does not work.
Did I miss something here?
Moreover, an example using the str() function with expressions would be very helpful 🙂
Thanks!

**Comment by aomarks** [comment, marked as answer]
> I'm trying to localize a property, using the str() function as specified in the doc:

The `str` tag is only needed if you need to put expressions inside a plain string. Otherwise you can directly pass a string to `msg`.

Valid:
```ts
msg('Hello world');
msg(str`Hello ${name}`)
msg(html`Hello <b>world</b>`);
msg(html`Hello <b>${name}</b>`);
```

Invalid:
<strike>

```ts
msg(`Hello ${name}`)
```

</strike>

(The `str` tag is needed when there are expressions because otherwise we receive a string where expressions are already interpolated into a plain string, so it wouldn't be possible to update the template with a translation and substitute a new expression into the placeholder).


> I would expect that my html label is translated when the lang is changed, but it does not work.

This is because `msg` returns a regular template or string in the current locale. The template/string itself doesn't know how to re-render itself when the locale changes, instead we rely on the `msg` call getting re-invoked each time the locale changes. The `@localized` decorator will automatically re-run `render` whenever the locale changes, so the key here is to ensure that any `msg` calls you have will be invoked whenever `render` runs.

One quick fix would be to move the default value directly into the `render` method:

```ts
@property({ reflect: true })
label?: string;

render() {
  return html`<label>${this.label ?? msg('Default label')}</label>`
}
```

Or if you prefer the property to handle the default in a more natural way, you can define your own getter/setter. This way, instead of the default being set once when the element is constructed, it is dynamic based on the current locale:

```ts
  private _label?: string;

  @property({ reflect: true })
  get label() {
    return this._label ?? msg('Default label');
  }

  set label(name: string) {
    this._label = name;
  }

  render() {
    return html`<label>${this.label}</label>`;
  }
```

> Moreover, an example using the str() function with expressions would be very helpful

We're currently working on new localize docs on lit.dev -- I'll make sure to clarify when to use `str`, and also to discuss the issue you ran into with templates not updating. By the way, the README at https://github.com/lit/lit/tree/main/packages/localize#litlocalize currently contains a fair bit more information than lit.dev does, if you haven't found that already! All of this extra information will be moving to lit.dev shortly.

Hope that helps!

**Reply by jeromecambon** [reply]
Thank you for the quick reply!

I've tried your proposal, but unfortunately it doesn't work for my need:
Initially the label property is null so at the first render() the label is initialized to 'Default label'.
When the locale is set to 'fr', the label is translated to 'Label par défaut'. So far so good.
But when I come back to the default locale ('en'), since the label has already a value, it is not translated back to 'Default label'...

Any idea on how I could make this work?
Thanks!

**Reply by aomarks** [reply]
> But when I come back to the default locale ('en'), since the label has already a value, it is not translated back to 'Default label'...

Odd! Can you share more of your code so I could take a look? Can you put a `console.log('render')` in your `render` method to check whether `render` is getting invoked when you switch back to the default locale?

[Here's an example](https://lit.dev/playground/#project=W3sibmFtZSI6IngtZ3JlZXRlci50cyIsImNvbnRlbnQiOiJpbXBvcnQge0xpdEVsZW1lbnQsIGh0bWx9IGZyb20gJ2xpdCc7XG5pbXBvcnQge2N1c3RvbUVsZW1lbnQsIHByb3BlcnR5fSBmcm9tICdsaXQvZGVjb3JhdG9ycy5qcyc7XG5pbXBvcnQge21zZywgbG9jYWxpemVkfSBmcm9tICdAbGl0L2xvY2FsaXplJztcblxuQGxvY2FsaXplZCgpXG5AY3VzdG9tRWxlbWVudCgneC1ncmVldGVyJylcbmV4cG9ydCBjbGFzcyBYR3JlZXRlciBleHRlbmRzIExpdEVsZW1lbnQge1xuICBAcHJvcGVydHkoKVxuICBncmVldGluZz86IHN0cmluZztcbiAgXG4gIHJlbmRlcigpIHtcbiAgICByZXR1cm4gaHRtbGA8cD4ke3RoaXMuZ3JlZXRpbmcgPz8gbXNnKGh0bWxgSGVsbG8gPGI-V29ybGQ8L2I-IWApfTwvcD5gO1xuICB9XG59XG4ifSx7Im5hbWUiOiJsb2NhbGUtcGlja2VyLnRzIiwiY29udGVudCI6Ii8qKlxuICogQGxpY2Vuc2VcbiAqIENvcHlyaWdodCAyMDIwIEdvb2dsZSBMTENcbiAqIFNQRFgtTGljZW5zZS1JZGVudGlmaWVyOiBCU0QtMy1DbGF1c2VcbiAqL1xuXG5pbXBvcnQge0xpdEVsZW1lbnQsIGh0bWx9IGZyb20gJ2xpdCc7XG5pbXBvcnQge2N1c3RvbUVsZW1lbnR9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJztcbmltcG9ydCB7Z2V0TG9jYWxlLCBzZXRMb2NhbGVGcm9tVXJsfSBmcm9tICcuL2xvY2FsaXphdGlvbi5qcyc7XG5pbXBvcnQge2FsbExvY2FsZXN9IGZyb20gJy4vbG9jYWxlLWNvZGVzLmpzJztcbmltcG9ydCB7bG9jYWxpemVkfSBmcm9tICdAbGl0L2xvY2FsaXplJztcblxuY29uc3QgbG9jYWxlTmFtZXM6IHtcbiAgW0wgaW4gdHlwZW9mIGFsbExvY2FsZXNbbnVtYmVyXV06IHN0cmluZztcbn0gPSB7XG4gIGVuOiAnRW5nbGlzaCcsXG4gICdlcy00MTknOiAnRXNwYcOxb2wgKExhdGlub2Ftw6lyaWNhKeKAjicsXG4gIHpoX0NOOiAn5Lit5paHICjnroDkvZMpJyxcbn07XG5cbi8vIE5vdGUgd2UgdXNlIHVwZGF0ZVdoZW5Mb2NhbGVDaGFuZ2VzIGhlcmUgc28gdGhhdCB3ZSdyZSBhbHdheXMgdXAgdG8gZGF0ZSB3aXRoXG4vLyB0aGUgYWN0aXZlIGxvY2FsZSAodGhlIHJlc3VsdCBvZiBnZXRMb2NhbGUoKSkgd2hlbiB0aGUgbG9jYWxlIGNoYW5nZXMgdmlhIGFcbi8vIGhpc3RvcnkgbmF2aWdhdGlvbi5cbkBsb2NhbGl6ZWQoKVxuQGN1c3RvbUVsZW1lbnQoJ2xvY2FsZS1waWNrZXInKVxuZXhwb3J0IGNsYXNzIExvY2FsZVBpY2tlciBleHRlbmRzIExpdEVsZW1lbnQge1xuICByZW5kZXIoKSB7XG4gICAgcmV0dXJuIGh0bWxgXG4gICAgICA8c2VsZWN0IEBjaGFuZ2U9JHt0aGlzLmxvY2FsZUNoYW5nZWR9PlxuICAgICAgICAke2FsbExvY2FsZXMubWFwKFxuICAgICAgICAgIChsb2NhbGUpID0-XG4gICAgICAgICAgICBodG1sYDxvcHRpb24gdmFsdWU9JHtsb2NhbGV9ID9zZWxlY3RlZD0ke2xvY2FsZSA9PT0gZ2V0TG9jYWxlKCl9PlxuICAgICAgICAgICAgICAke2xvY2FsZU5hbWVzW2xvY2FsZV19XG4gICAgICAgICAgICA8L29wdGlvbj5gXG4gICAgICAgICl9XG4gICAgICA8L3NlbGVjdD5cbiAgICBgO1xuICB9XG5cbiAgbG9jYWxlQ2hhbmdlZChldmVudDogRXZlbnQpIHtcbiAgICBjb25zdCBuZXdMb2NhbGUgPSAoZXZlbnQudGFyZ2V0IGFzIEhUTUxTZWxlY3RFbGVtZW50KS52YWx1ZTtcbiAgICBpZiAobmV3TG9jYWxlICE9PSBnZXRMb2NhbGUoKSkge1xuICAgICAgY29uc3QgdXJsID0gbmV3IFVSTCh3aW5kb3cubG9jYXRpb24uaHJlZik7XG4gICAgICB1cmwuc2VhcmNoUGFyYW1zLnNldCgnbG9jYWxlJywgbmV3TG9jYWxlKTtcbiAgICAgIHdpbmRvdy5oaXN0b3J5LnB1c2hTdGF0ZShudWxsLCAnJywgdXJsLnRvU3RyaW5nKCkpO1xuICAgICAgc2V0TG9jYWxlRnJvbVVybCgpO1xuICAgIH1cbiAgfVxufVxuIn0seyJuYW1lIjoiaW5kZXgudHMiLCJjb250ZW50IjoiLyoqXG4gKiBAbGljZW5zZVxuICogQ29weXJpZ2h0IDIwMjAgR29vZ2xlIExMQ1xuICogU1BEWC1MaWNlbnNlLUlkZW50aWZpZXI6IEJTRC0zLUNsYXVzZVxuICovXG5cbmltcG9ydCB7c2V0TG9jYWxlRnJvbVVybH0gZnJvbSAnLi9sb2NhbGl6YXRpb24uanMnO1xuaW1wb3J0IHtMT0NBTEVfU1RBVFVTX0VWRU5UfSBmcm9tICdAbGl0L2xvY2FsaXplJztcbmltcG9ydCB7aHRtbCwgcmVuZGVyfSBmcm9tICdsaXQnO1xuaW1wb3J0ICcuL2xvY2FsZS1waWNrZXIuanMnO1xuaW1wb3J0ICcuL3gtZ3JlZXRlci5qcyc7XG5pbXBvcnQgJ0BtYXRlcmlhbC9td2MtY2lyY3VsYXItcHJvZ3Jlc3MnO1xuXG5jb25zdCBtYWluID0gZG9jdW1lbnQucXVlcnlTZWxlY3RvcignbWFpbicpITtcbmNvbnN0IHNwaW5uZXIgPSBkb2N1bWVudC5xdWVyeVNlbGVjdG9yKCcjc3Bpbm5lcicpITtcblxuLy8gVXBkYXRlIHRoZSBsb2NhbGUgdG8gbWF0Y2ggdGhlIFVSTCB3aGVuIHRoZSB1c2VyIG1vdmVzIGJhY2t3YXJkcyBvciBmb3J3YXJkc1xuLy8gdGhyb3VnaCBoaXN0b3J5Llxud2luZG93LmFkZEV2ZW50TGlzdGVuZXIoJ3BvcHN0YXRlJywgKCkgPT4ge1xuICBzZXRMb2NhbGVGcm9tVXJsKCk7XG59KTtcblxuLy8gRGlzcGxheSBhIHNwaW5uZXIgd2hlbmV2ZXIgYSBuZXcgbG9jYWxlIGlzIGxvYWRpbmcuXG53aW5kb3cuYWRkRXZlbnRMaXN0ZW5lcihMT0NBTEVfU1RBVFVTX0VWRU5ULCAoe2RldGFpbH0pID0-IHtcbiAgaWYgKGRldGFpbC5zdGF0dXMgPT09ICdsb2FkaW5nJykge1xuICAgIGNvbnNvbGUubG9nKGBMb2FkaW5nIG5ldyBsb2NhbGU6ICR7ZGV0YWlsLmxvYWRpbmdMb2NhbGV9YCk7XG4gICAgc3Bpbm5lci5yZW1vdmVBdHRyaWJ1dGUoJ2hpZGRlbicpO1xuICB9IGVsc2UgaWYgKGRldGFpbC5zdGF0dXMgPT09ICdyZWFkeScpIHtcbiAgICBjb25zb2xlLmxvZyhgTG9hZGVkIG5ldyBsb2NhbGU6ICR7ZGV0YWlsLnJlYWR5TG9jYWxlfWApO1xuICAgIHNwaW5uZXIuc2V0QXR0cmlidXRlKCdoaWRkZW4nLCAnJyk7XG4gIH0gZWxzZSBpZiAoZGV0YWlsLnN0YXR1cyA9PT0gJ2Vycm9yJykge1xuICAgIGNvbnNvbGUuZXJyb3IoXG4gICAgICBgRXJyb3IgbG9hZGluZyBsb2NhbGUgJHtkZXRhaWwuZXJyb3JMb2NhbGV9OiBgICsgZGV0YWlsLmVycm9yTWVzc2FnZVxuICAgICk7XG4gICAgc3Bpbm5lci5zZXRBdHRyaWJ1dGUoJ2hpZGRlbicsICcnKTtcbiAgfVxufSk7XG5cbihhc3luYyAoKSA9PiB7XG4gIHRyeSB7XG4gICAgLy8gRGVmZXIgZmlyc3QgcmVuZGVyIHVudGlsIG91ciBpbml0aWFsIGxvY2FsZSBpcyByZWFkeSwgdG8gYXZvaWQgYSBmbGFzaCBvZlxuICAgIC8vIHRoZSB3cm9uZyBsb2NhbGUuXG4gICAgYXdhaXQgc2V0TG9jYWxlRnJvbVVybCgpO1xuICB9IGNhdGNoIChlKSB7XG4gICAgLy8gRWl0aGVyIHRoZSBVUkwgbG9jYWxlIGNvZGUgd2FzIGludmFsaWQsIG9yIHRoZXJlIHdhcyBhIHByb2JsZW0gbG9hZGluZ1xuICAgIC8vIHRoZSBsb2NhbGUgbW9kdWxlLlxuICAgIGNvbnNvbGUuZXJyb3IoYEVycm9yIGxvYWRpbmcgbG9jYWxlOiAkeyhlIGFzIEVycm9yKS5tZXNzYWdlfWApO1xuICB9XG4gIHJlbmRlcihodG1sYFxuICAgICAgIDx4LWdyZWV0ZXI-PC94LWdyZWV0ZXI-XG4gICAgICAgPHgtZ3JlZXRlciBncmVldGluZz1cIkN1c3RvbSBncmVldGluZ1wiPjwveC1ncmVldGVyPlxuICAgICAgICAgYCwgbWFpbik7XG59KSgpO1xuIn0seyJuYW1lIjoiaW5kZXguaHRtbCIsImNvbnRlbnQiOiI8IURPQ1RZUEUgaHRtbD5cbjxoZWFkPlxuICA8dGl0bGU-bGl0LWxvY2FsaXplIHJ1bnRpbWUgZXhhbXBsZTwvdGl0bGU-XG4gIDxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vaW5kZXguanNcIj48L3NjcmlwdD5cbiAgPHN0eWxlPlxuICAgIFtoaWRkZW5dIHtcbiAgICAgIHZpc2liaWxpdHk6IGhpZGRlbjtcbiAgICB9XG4gICAgI3NwaW5uZXIge1xuICAgICAgcG9zaXRpb246IGFic29sdXRlO1xuICAgIH1cbiAgPC9zdHlsZT5cbjwvaGVhZD5cbjxib2R5PlxuICA8bG9jYWxlLXBpY2tlcj48L2xvY2FsZS1waWNrZXI-XG4gIDxtd2MtY2lyY3VsYXItcHJvZ3Jlc3NcbiAgICBpZD1cInNwaW5uZXJcIlxuICAgIGluZGV0ZXJtaW5hdGVcbiAgICBoaWRkZW5cbiAgPjwvbXdjLWNpcmN1bGFyLXByb2dyZXNzPlxuICA8bWFpbj48L21haW4-XG48L2JvZHk-XG4ifSx7Im5hbWUiOiJsb2NhbGVzL2VzLTQxOS50cyIsImNvbnRlbnQiOiIvLyBEbyBub3QgbW9kaWZ5IHRoaXMgZmlsZSBieSBoYW5kIVxuLy8gUmUtZ2VuZXJhdGUgdGhpcyBmaWxlIGJ5IHJ1bm5pbmcgbGl0LWxvY2FsaXplXG5cbmltcG9ydCB7aHRtbH0gZnJvbSAnbGl0JztcblxuLyogZXNsaW50LWRpc2FibGUgbm8taXJyZWd1bGFyLXdoaXRlc3BhY2UgKi9cbi8qIGVzbGludC1kaXNhYmxlIEB0eXBlc2NyaXB0LWVzbGludC9uby1leHBsaWNpdC1hbnkgKi9cblxuZXhwb3J0IGNvbnN0IHRlbXBsYXRlcyA9IHtcbiAgaDNjNDRhZmYyZDVmNWVmNmI6IGh0bWxgSG9sYSA8Yj5NdW5kbzwvYj4hYCxcbn07XG4ifSx7Im5hbWUiOiJsb2NhbGVzL3poX0NOLnRzIiwiY29udGVudCI6Ii8vIERvIG5vdCBtb2RpZnkgdGhpcyBmaWxlIGJ5IGhhbmQhXG4vLyBSZS1nZW5lcmF0ZSB0aGlzIGZpbGUgYnkgcnVubmluZyBsaXQtbG9jYWxpemVcblxuaW1wb3J0IHtodG1sfSBmcm9tICdsaXQnO1xuXG4vKiBlc2xpbnQtZGlzYWJsZSBuby1pcnJlZ3VsYXItd2hpdGVzcGFjZSAqL1xuLyogZXNsaW50LWRpc2FibGUgQHR5cGVzY3JpcHQtZXNsaW50L25vLWV4cGxpY2l0LWFueSAqL1xuXG5leHBvcnQgY29uc3QgdGVtcGxhdGVzID0ge1xuICBoM2M0NGFmZjJkNWY1ZWY2YjogaHRtbGDkvaDlpb0sIDxiPuS4lueVjCE8L2I-YCxcbn07XG4ifSx7Im5hbWUiOiJsb2NhbGl6YXRpb24udHMiLCJjb250ZW50IjoiaW1wb3J0IHtjb25maWd1cmVMb2NhbGl6YXRpb259IGZyb20gJ0BsaXQvbG9jYWxpemUnO1xuaW1wb3J0IHtzb3VyY2VMb2NhbGUsIHRhcmdldExvY2FsZXN9IGZyb20gJy4vbG9jYWxlLWNvZGVzLmpzJztcblxuZXhwb3J0IGNvbnN0IHtnZXRMb2NhbGUsIHNldExvY2FsZX0gPSBjb25maWd1cmVMb2NhbGl6YXRpb24oe1xuICBzb3VyY2VMb2NhbGUsXG4gIHRhcmdldExvY2FsZXMsXG4gIGxvYWRMb2NhbGU6IChsb2NhbGU6IHN0cmluZykgPT4gaW1wb3J0KGAuL2xvY2FsZXMvJHtsb2NhbGV9LmpzYCksXG59KTtcblxuZXhwb3J0IGNvbnN0IHNldExvY2FsZUZyb21VcmwgPSBhc3luYyAoKSA9PiB7XG4gIGNvbnN0IHVybCA9IG5ldyBVUkwod2luZG93LmxvY2F0aW9uLmhyZWYpO1xuICBjb25zdCBsb2NhbGUgPSB1cmwuc2VhcmNoUGFyYW1zLmdldCgnbG9jYWxlJykgfHwgc291cmNlTG9jYWxlO1xuICBhd2FpdCBzZXRMb2NhbGUobG9jYWxlKTtcbn07XG4ifSx7Im5hbWUiOiJsb2NhbGUtY29kZXMudHMiLCJjb250ZW50IjoiLy8gRG8gbm90IG1vZGlmeSB0aGlzIGZpbGUgYnkgaGFuZCFcbi8vIFJlLWdlbmVyYXRlIHRoaXMgZmlsZSBieSBydW5uaW5nIGxpdC1sb2NhbGl6ZS5cblxuLyoqXG4gKiBUaGUgbG9jYWxlIGNvZGUgdGhhdCB0ZW1wbGF0ZXMgaW4gdGhpcyBzb3VyY2UgY29kZSBhcmUgd3JpdHRlbiBpbi5cbiAqL1xuZXhwb3J0IGNvbnN0IHNvdXJjZUxvY2FsZSA9IGBlbmA7XG5cbi8qKlxuICogVGhlIG90aGVyIGxvY2FsZSBjb2RlcyB0aGF0IHRoaXMgYXBwbGljYXRpb24gaXMgbG9jYWxpemVkIGludG8uIFNvcnRlZFxuICogbGV4aWNvZ3JhcGhpY2FsbHkuXG4gKi9cbmV4cG9ydCBjb25zdCB0YXJnZXRMb2NhbGVzID0gW2Blcy00MTlgLCBgemhfQ05gXSBhcyBjb25zdDtcblxuLyoqXG4gKiBBbGwgdmFsaWQgcHJvamVjdCBsb2NhbGUgY29kZXMuIFNvcnRlZCBsZXhpY29ncmFwaGljYWxseS5cbiAqL1xuZXhwb3J0IGNvbnN0IGFsbExvY2FsZXMgPSBbYGVuYCwgYGVzLTQxOWAsIGB6aF9DTmBdIGFzIGNvbnN0O1xuIn0seyJuYW1lIjoicGFja2FnZS5qc29uIiwiY29udGVudCI6IntcbiAgXCJkZXBlbmRlbmNpZXNcIjoge1xuICAgIFwibGl0XCI6IFwiXjIuMC4wXCIsXG4gICAgXCJAbGl0L3JlYWN0aXZlLWVsZW1lbnRcIjogXCJeMS4wLjBcIixcbiAgICBcImxpdC1lbGVtZW50XCI6IFwiXjMuMC4wXCIsXG4gICAgXCJsaXQtaHRtbFwiOiBcIl4yLjAuMFwiXG4gIH1cbn0iLCJoaWRkZW4iOnRydWV9XQ) where I modified the localization example from our docs to show how you could fall back to a default that uses `msg`. Switching away from and back to the default locale is working here.

**Reply by jeromecambon** [reply]
ok, I've fixed my issue: I was checking 'targetLocales' instead of 'allLocales' for locale update. So I was unable to change back to 'en'.

Thank you very much for your help et responsiveness! 😃
